src/api/process.c

```diff
@@ -63,60 +84,210 @@ typedef enum {
+  SDL_DestroyCond(list->work_done);
+  node = list->head;
+  while (node) {
+    temp = node;
+    node = node->next;
+    free(temp);
+  }
+}
+
+
+static bool kill_list_init(process_kill_list_t *list) {
+  list->size = 0;
+  list->mutex = SDL_CreateMutex();
+  list->has_work = SDL_CreateCond();
+  list->work_done = SDL_CreateCond();
+  list->head = list->tail = NULL;
+  list->stop = false;
+  if (!list->mutex || !list->has_work || !list->work_done) {
+    kill_list_free(list);
+    return false;
+  }
+  return true;
+}
+
+
+static void kill_list_push(process_kill_list_t *list, process_kill_t *task) {
+  if (!list) return;
+  task->next = NULL;
+  if (list->tail) {
+    list->tail->next = task;
+    list->tail = task;
+  } else {
+    list->head = list->tail = task;
+  }
+  list->size++;
+}
+
+
+static void kill_list_pop(process_kill_list_t *list) {
+  if (!list || !list->head) return;
+  process_kill_t *head = list->head;
+  list->head = list->head->next;
+  if (!list->head) list->tail = NULL;
+  head->next = NULL;
+  list->size--;
+}
+
+
+static void kill_list_wait_all(process_kill_list_t *list) {
+  SDL_LockMutex(list->mutex);
+  // wait until list is empty
+  while (list->size)
+    SDL_CondWait(list->work_done, list->mutex);
+  // tell the worker to stop
+  list->stop = true;
+  SDL_CondSignal(list->has_work);
+  SDL_UnlockMutex(list->mutex);
+}
+
+
+static void process_handle_close(process_handle *handle) {
+#ifdef _WIN32
+  if (*handle) {
+    CloseHandle(*handle);
+    *handle = NULL;
+  }
 #endif
+  (void) 0;
+}
+
+
+static bool process_handle_is_running(process_handle handle, int *status) {
+#ifdef _WIN32
+  DWORD s;
+  if (GetExitCodeProcess(handle, &s) && s != STILL_ACTIVE) {
+    if (status != NULL)
+      *status = s;
+    return false;
+  }
+#else
+  int s;
+  if (waitpid(handle, &s, WNOHANG) != 0) {
+    if (status != NULL)
+      *status = WEXITSTATUS(s);
+    return false;
+  }
+#endif
+  return true;
+}
+
+
+static bool process_handle_signal(process_handle handle, signal_e sig) {
+#if _WIN32
+  switch(sig) {
+    case SIGNAL_TERM: return GenerateConsoleCtrlEvent(CTRL_BREAK_EVENT, GetProcessId(handle));
+    case SIGNAL_KILL: return TerminateProcess(handle, -1);
+    case SIGNAL_INTERRUPT: return DebugBreakProcess(handle);
+  }
+#else
+  switch (sig) {
+    case SIGNAL_TERM: return kill(-handle, SIGTERM) == 0; break;
+    case SIGNAL_KILL: return kill(-handle, SIGKILL) == 0; break;
+    case SIGNAL_INTERRUPT: return kill(-handle, SIGINT) == 0; break;
+  }
+#endif
+  return false;
+}
+
+
+static int kill_list_worker(void *ud) {
+  process_kill_list_t *list = (process_kill_list_t *) ud;
+  process_kill_t *current_task;
+  int i, size;
+
+  while (true) {
+    SDL_LockMutex(list->mutex);
+
+    // wait until we have work to do
+    while (!list->head && !list->stop)
+      SDL_CondWait(list->has_work, list->mutex); // LOCK MUTEX
+
+    if (list->stop) break;
+
+    for (i = 0, size = list->size; i < size; i++) {
```

Comment:
Instead of this `for`, could we use something like `while (current_task)`?
Just a nit, but I feel like it would be a bit cleaner this way.

Reply:
The whole reason size was added into the struct was because you can't simply use `while (current_task)` if you want to push to the same list. It has potential to deadlock (or at least spinlock until all tasks completed).

I've considered tracking list->tail and checking that, but in the end settled for this solution because it is the simplest to understand and implement.